pkg/app/pipedv1/plugin/kubernetes/provider/deployment.go

```diff
@@ -0,0 +1,129 @@
+// - spec.template.spec.containers.env.valueFrom.configMapKeyRef.name
+// - spec.template.spec.containers.envFrom.configMapRef.name
+func FindReferencingConfigMaps(m *unstructured.Unstructured) []string {
+	var configMaps []string
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "volumes", "configMap", "name"); len(n) > 0 {
+		configMaps = append(configMaps, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "initContainers", "env", "valueFrom", "configMapKeyRef", "name"); len(n) > 0 {
+		configMaps = append(configMaps, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "initContainers", "envFrom", "configMapRef", "name"); len(n) > 0 {
+		configMaps = append(configMaps, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "containers", "env", "valueFrom", "configMapKeyRef", "name"); len(n) > 0 {
+		configMaps = append(configMaps, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "containers", "envFrom", "configMapRef", "name"); len(n) > 0 {
+		configMaps = append(configMaps, n...)
+	}
+
+	slices.Sort(configMaps)
+	return slices.Compact(configMaps)
+}
+
+// FindReferencingSecrets finds all secrets that are referenced by the given manifest.
+//
+// It looks for secrets in the following fields:
+// - spec.template.spec.volumes.secret.secretName
+// - spec.template.spec.initContainers.env.valueFrom.secretKeyRef.name
+// - spec.template.spec.initContainers.envFrom.secretRef.name
+// - spec.template.spec.containers.env.valueFrom.secretKeyRef.name
+// - spec.template.spec.containers.envFrom.secretRef.name
+func FindReferencingSecrets(m *unstructured.Unstructured) []string {
+	var secrets []string
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "volumes", "secret", "secretName"); len(n) > 0 {
+		secrets = append(secrets, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "initContainers", "env", "valueFrom", "secretKeyRef", "name"); len(n) > 0 {
+		secrets = append(secrets, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "initContainers", "envFrom", "secretRef", "name"); len(n) > 0 {
+		secrets = append(secrets, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "containers", "env", "valueFrom", "secretKeyRef", "name"); len(n) > 0 {
+		secrets = append(secrets, n...)
+	}
+
+	if n := nestedStringSlice(m.Object, "spec", "template", "spec", "containers", "envFrom", "secretRef", "name"); len(n) > 0 {
+		secrets = append(secrets, n...)
+	}
+
+	slices.Sort(secrets)
+	return slices.Compact(secrets)
+}
+
+// nestedStringSlice extracts a string slice from the given object by following the fields.
+// It returns the extracted string slice.
+// If there is []map[string]any in the middle of the fields, it will be flattened.
+func nestedStringSlice(obj any, fields ...string) []string {
```

Comment:
I don't want to convert the unstructured.Unstructured to appsv1.Deployment or something, so I implemented this.
We cannot use `unstructured.NestedStringSlice` because it handles the `[]map[string]any` differently.

Reply:
@Warashi 
[Ask] 
> We cannot use unstructured.NestedStringSlice because it handles the []map[string]any differently.

What's the difference between `unstructured.NestedStringSlice` and `nestedStringSlice` ?

Reply:
@ffjlabo 
`unstructured.NestedStringSlice` returns an error when there is `[]map[string]any` in the middle of the fields. On the other hand, `nestedStringSlice` flattens it.

Reply:
@Warashi 
I got your point. I tried to use `unstructured.NestedStringSlice` with the data below.
We can't find and flatten slices in the `[]interface{}{map[string]interface{}{}}`.

```
	obj := map[string]interface{}{
		"spec": map[string]interface{}{
			"template": map[string]interface{}{
				"spec": map[string]interface{}{
					"volumes": []interface{}{
						map[string]interface{}{
							"configMap": map[string]interface{}{
								"name": "canary-by-config-change",
							},
						},
					},
				},
			},
		},
	}

	res, exists, err := unstructured.NestedStringSlice(obj, "spec", "template", "spec", "volumes[]", "configMap", "name")
	assert.True(t, exists)
	assert.NoError(t, err)
	assert.Equal(t, res, []string{"canary-by-config-change"})
```

```
=== RUN   Test
        	Error:      	Should be true
        	Error:      	Received unexpected error:
        	            	.spec.template.spec.volumes.configMap accessor error: [map[configMap:map[name:canary-by-config-change]]] is of the type []interface {}, expected map[string]interface{}
        	Error:      	Not equal:
        	            	expected: []string(nil)
        	            	actual  : []string{"canary-by-config-change"}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,4 @@
        	            	-([]string) <nil>
        	            	+([]string) (len=1) {
        	            	+ (string) (len=23) "canary-by-config-change"
        	            	+}

        	Test:       	Test
```

The test for `unstructured.NestedFieldNoCopy` (used in the `unstructured.NestedStringSlice`) shows it.
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/unstructured/helpers_test.go#L122